rust/lance/src/dataset/write/merge_insert.rs

```diff
@@ -993,13 +1008,41 @@ impl MergeInsertJob {
     /// This will take in the source, merge it with the existing target data, and insert new
     /// rows, update existing rows, and delete existing rows
     pub async fn execute(
-        self,
+        mut self,
         source: SendableRecordBatchStream,
     ) -> Result<(Arc<Dataset>, MergeStats)> {
-        let ds = self.dataset.clone();
-        let (transaction, stats) = self.execute_uncommitted_impl(source).await?;
-        let dataset = CommitBuilder::new(ds).execute(transaction).await?;
-        Ok((Arc::new(dataset), stats))
+        let mut source_iter =
+            super::new_source_iter(source, self.params.conflict_retries > 0).await?;
+
+        let mut dataset_ref = self.dataset.clone();
+        let max_retries = self.params.conflict_retries;
+        let mut backoff = Backoff::default();
+        while backoff.attempt() <= max_retries {
+            let ds = dataset_ref.clone();
+            let (transaction, stats) = self
+                .clone()
+                .execute_uncommitted_impl(source_iter.next().unwrap())
+                .await?;
+            match CommitBuilder::new(ds).execute(transaction).await {
+                Ok(ds) => return Ok((Arc::new(ds), stats)),
+                Err(Error::RetryableCommitConflict { .. }) => {
+                    tokio::time::sleep(backoff.next_backoff()).await;
+                    let mut ds = dataset_ref.as_ref().clone();
+                    ds.checkout_latest().await?;
+                    dataset_ref = Arc::new(ds);
+                    self.dataset = dataset_ref.clone();
+                    continue;
+                }
+                Err(e) => return Err(e),
+            };
+        }
+        Err(Error::TooMuchContention {
+            message: format!(
+                "Attempted {} retries, but there is too much contention on the table.",
```

Comment:
"contention" is the most succinct term but I wonder if it might be a bit too ambiguous for users. 

Something like "too many concurrent writes" might be a bit easier to understand